### PR TITLE
Added ConditionPromise.

### PR DESCRIPTION
**Pull request by emilwestergren**
@erikhagglund peer review

### CODE REVIEW

**Comment by marcusnaslund** [comment]
Add test plz

**Comment by emilwestergren** [comment]
Added test

File: source/concurrent/Promise.ooc [hunk 1]
```diff
@@ -61,6 +61,30 @@ ClosurePromise: class extends Promise {
...
+		(this _mutex, this _condition) free()
+		super()
+	}
+	signal: func {
+		this _mutex with(||
+			this _completed = true
+			this _condition broadcast()
+		)
+	}
+	wait: override func (time: TimeSpan) -> Bool {
```
Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: I like how this function takes a time argument but never uses it.

**Comment by marcusnaslund** [comment]
...but since I locked you in the office yesterday, I have no comments.

File: source/concurrent/Promise.ooc (hunk 1 above)
Comment (emilwestergren) [inline source/concurrent/Promise.ooc]: Yeah, I guess I could do it like this instead:
```ooc
wait: override func (time: TimeSpan) -> Bool { Debug error("Timeout wait not supported for ConditionPromise") }
wait: override func -> Bool { ... }
```

Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: That would make more sense, but it would override the `debugDeadlock` code. You decide.

Comment (emilwestergren) [inline source/concurrent/Promise.ooc]: How about this:
```ooc
if (time != TimeSpan maximumValue)
   Debug error("Timed wait not supported...")
```

Comment (emilwestergren) [inline source/concurrent/Promise.ooc]: That won't work with `debugDeadlock` though...

Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: the timed wait should give an error if the argument is not `TimeSpan maximumValue`, but keep all the code you have now.
Then you can always call `wait()` without argument, and everyone is happy.

**Comment by emilwestergren** [comment]
Updated

File: source/concurrent/Promise.ooc (hunk 1 above)
Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: Ah, true

**Pull request merged**